src/shared/utils/guard.ts

```diff
@@ -0,0 +1,43 @@
+import { SchemaId, SignedWrappedDocument } from "../..";
+import {
+  OpenAttestationDocument as OpenAttestationDocumentV3,
+  VerifiableCredentialProofBase,
+  VerifiableCredentialProofSigned,
+  WrappedDocument as WrappedDocumentV3
+} from "../../3.0/types";
+import {
+  ArrayProof,
+  OpenAttestationDocument as OpenAttestationDocumentV2,
+  Signature,
+  WrappedDocument as WrappedDocumentV2
+} from "../../2.0/types";
+import { getSchema, validateSchema as validate } from "../validate";
+
+export const isWrappedV3Document = (document: any): document is WrappedDocumentV3<OpenAttestationDocumentV3> => {
+  if (!document || typeof document !== "object" || document.version !== SchemaId.v3) return false;
+  const validateSchema = () => validate(document, getSchema(SchemaId.v3)).length === 0;
+  const validateProof = () => VerifiableCredentialProofBase.guard(document.proof);
+  return validateSchema() && validateProof();
+};
+
+export const isWrappedV2Document = (document: any): document is WrappedDocumentV2<OpenAttestationDocumentV2> => {
+  if (!document || typeof document !== "object") return false;
+  const validateSchema = () => validate(document, getSchema(SchemaId.v2)).length === 0;
+  const validateSignature = () => Signature.guard(document.signature);
+  return validateSchema() && validateSignature();
+};
+
+export const isSignedWrappedV2Document = (
+  document: any
+): document is SignedWrappedDocument<OpenAttestationDocumentV2> => {
+  const validateWrappedDocument = () => isWrappedV2Document(document);
```

Comment:
why do you write it as a closure?

Reply:
there were a reason (actually still). After @cavacado to split the verification into parts, I had issues. I had to check the document is not null or undefined, otherwise as the evaluation happens before the conditional, it will throw an error at `document.proof`.

It happened as well in SignedWrappedDocument.

WrappedDocument guard are fine now (because I moved the check at the top), but not SignedWrappedDocument (unless I add another check at the top ... :) )

So in order to make it consistent everywhere I just decided to wrap into function (lazy evaluation)